### PR TITLE
Fix building with MSYS2 distributed MinGW hdf5.

### PR DESCRIPTION
**Pull request by Berrysoft**
* Allow minor version to be 14, because the version of MSYS2 distributed MinGW hdf5 is 1.14. Maybe need: #227 
* Add `is_msvc` to `build.rs` to determine target enviroment from env var instead of `cfg!`, to make cross-compile right. Here the term "cross-compile" also means targeting GNU with MSVC toolchain (which is very common).
* The dll name of hdf5 is `libhdf5-0.dll` or `libhdf5-310.dll` (see this commit: https://github.com/msys2/MINGW-packages/commit/205f329c0231d8046b8bb3a2f6dff787c6613347). I'm afraid we need to update the dll name every time MSYS2 updates it. I suggest using `pkgconf` to detect it in the future.
* Limit `msvc_dll_indirection` to MSVC only.
* The tests have passed on my machine with MSVC toolchain, targeting GNU, and latest MSYS2.

### CODE REVIEW

**Comment by mulimoen** [comment]
Do you have a way of including this build in the CI?

**Comment by Berrysoft** [comment]
I will try to include it. And by the way, `cargo-clippy` fails because of `uninlined_format_args`. Maybe you need to turn it off because it's very annoying:)

**Comment by Berrysoft** [comment]
@mulimoen I've added a job to test this crate with MSYS2.

**Comment by mulimoen** [comment]
The clippy lints should be fixed in a separate PR, I think clippy supports `rust_version` which would prevent MSRV trouble in the future. We need a bump in MSRV for that...

**Comment by Berrysoft** [comment]
The package updated just now, and the dll name now changes with the soversion. Should we use `pkgconf` in that case, instead of hard-coding the dll name manually?

**Comment by Berrysoft** [comment]
The new test passed.

**Comment by mulimoen** [comment]
I have not used a windows system in ages. I think you are a better judge of whether `pkgconf` will be more or less suprising for the developers using this library

**Comment by Berrysoft** [comment]
MSYS2 provides `pkgconf` of course, and `vcpkg`, the Microsoft-banded C/C++ package manager, [also provides](https://github.com/microsoft/vcpkg/tree/master/ports/pkgconf) `pkgconf` affairs. I have also noticed that some other bindings, for example, [libgit2](https://github.com/rust-lang/git2-rs/blob/master/libgit2-sys/build.rs) and [openssl](https://github.com/sfackler/rust-openssl/blob/master/openssl-sys/build/find_normal.rs) also use `pkgconf` to find libs.

`openssl-sys` only uses `pkgconf` for Windows GNU target. I think I can follow this approach.

**Comment by Berrysoft** [comment]
I added pkg-config support for all windows target because the gnu and msvc targets all supports pkg-config, and (maybe a bug?) they cannot be dealt separately. (When targeting GNU, a MSVC target build script will also be built and executed.)

And for Windows target, I choose to trust pkg-config result and not checking the dll, because at first we wanted to avoid hard-coding the dll names with pkg-config, which could not provide the dll name either.

**Comment by mulimoen** [comment]
We know hdf5 localisation is difficult, especially on windows. Thanks a lot for doing this the proper way

**Comment by Berrysoft** [comment]
Pkg-config for Windows support added. Now it should work with pkg-config, and if pkg-config is not installed, `HDF5_DIR` would also work (if dll name hadn't been changed by MSYS2).

`pkg-config` gives many paths like `C:\msys64\mingw64\lib\..\bin`, so we need to canonicalize the path before querying it in `PATH`. The paths in `PATH` should also be canonicalized when comparing because `canonicalize` would turn a path to a UNC path on Windows.

**Review by mulimoen — CHANGES_REQUESTED** [review]
Just some minor comments, otherwise this PR is in great shape. Would you mind adding to the changelog?

File: hdf5-sys/build.rs
Comment (Berrysoft) [reply hdf5-sys/build.rs:32]: I'm waiting for #227 merge.

File: hdf5-sys/build.rs
Comment (Berrysoft) [reply hdf5-sys/build.rs:258]: Reference: https://github.com/sfackler/rust-openssl/blob/master/openssl-sys/build/find_normal.rs#L205-L207

At least openssl-sys doesn't reset this variable:)

This variable may cause some mismatching when there are many `pkg-config` installed on one system in particular cases (for example, MinGW `pkg-config` is called when using MSVC toolchain), but the user can avoid that by setting `PKG_CONFIG_PATH` manually.

Comment (mulimoen) [reply hdf5-sys/build.rs:258]: Great, thanks for clarifying!

**Review by mulimoen — APPROVED** [review]

**Comment by aldanor** [comment]
If there's still interest in moving this forward, it would have to be rebased on the latest master, now that 1.14 support is in and all the build failures etc have been hopefully resolved and fixed.

**Comment by Berrysoft** [comment]
I've merged master but need approval to run CI.

**Comment by aldanor** [comment]
@Berrysoft There's changelog conflicts so you might want to quickly rebase again.

It looks like your build ran on CI and failed: https://github.com/aldanor/hdf5-rust/actions/runs/5227987760/jobs/9440874079?pr=229

Because the version is `1.14.1.2` (I think we do parse `1.14.1-2`, like it looks on arch, but not this format)

**Comment by Berrysoft** [comment]
I solved the conflicts and updated the regex. Now it should parse both kind of versions.

**Comment by aldanor** [comment]
@mulimoen Looks good? Any comments from your side? Seems like mingw build now runs green again 1.14.1.2.

// @Berrysoft not to be picky, minor nit: in PRs you'd normally _rebase_ your branch against the main one as opposed to merging, otherwise you end up with all those merge commits in main. No big deal though, I'll just squash-merge it.

**Comment by aldanor** [comment]
Interestingly enough, this mingw run doesn't hit this problem https://github.com/HDFGroup/hdf5/issues/3091

**Comment by Berrysoft** [comment]
I cannot reproduce this problem on my device either.

**Comment by mulimoen** [comment]
Weird with the 1.14.1-2 static build, can't see any obvious differences in the build flags. Looks good to go!

**Comment by aldanor** [comment]
Now merged, thanks @Berrysoft

**Pull request merged**
